docs/source/meta_init.rst

```diff
@@ -0,0 +1,239 @@
+its underlying machinery is actually more flexible. You can also construct and
+later materialize individual tensors.
+
+::
+
+    >>> with torch.nn.utils.meta_init():
+    ...     t1 = torch.ones([3])
+    ...     t2 = t1 + 2
+    >>> t1
+    tensor(..., device='meta', size=(3,))
+    >>> t2
+    tensor(..., device='meta', size=(3,))
+    >>> # By default, `materialize()` clears the meta-init cache. If you want
+    ... # to materialize more than one object, set the `keep_cache` parameter
+    ... # to `True`.
+    >>> torch.nn.utils.materialize(t1, keep_cache=True)
+    >>> t1
+    tensor([1., 1., 1.])
+    >>> torch.nn.utils.materialize(t2)
+    >>> t2
+    tensor([3., 3., 3.])
+
+
+API
+---
+.. autosummary::
+    :toctree: generated
+    :nosignatures:
+
+    meta_init
+    is_meta_init
+    materialize
+    clear_meta_init_cache
+
+.. autofunction:: meta_init
+.. autofunction:: is_meta_init
+.. autofunction:: materialize
+.. autofunction:: clear_meta_init_cache
+
+.. _common-failures:
+
+Common Failure Cases
+--------------------
+Since meta initialization relies on constructing modules and their tensors on
+the meta device, there are some certain usage patterns that will fail when used
+with :func:`meta_init`.
+
+**A module that uses an operator not supported by the meta backend:**
+If the module uses an operator that is not yet supported by the meta backend,
+the operator call will fail. If you experience such failure, please open a
+GitHub issue and let us know which operator you are using.
+
+**A module that has device-specific initialization logic:**
+A module that expects to be initialized using one or more specific device types
+and includes initialization logic such as:
+
+::
+
+  def MyModule(Module):
+      def __init__(self):
+          x = torch.ones([10, 10])
+          if x.is_cpu():
+              y = x + 1
+          elif x.is_cuda():
+              y = x + 2
+
+
+will silently fail, potentially with some partial initialization.
+
+**A module that uses auxiliary state beyond its parameters and buffers:**
+The :func:`materialize` function simply traverses through the parameters and
+buffers of a module and materializes them. This means if the module has some
+other auxiliary state those tensors won't be materialized.
+
+**A module that directly accesses to raw tensor storage:**
+If the module attempts to directly read from or write to the raw storage of a
+tensor using a naked pointer without checking its validity, it will cause a
+segmentation fault since meta tensors don't have any storage allocated to them.
+
+**A tensor loaded from external data:**
+Although not necessarily a failure, if a tensor is constructed from external
+data (e.g. numpy), :func:`meta_init` won't intercept it and will allow the
+tensor to be constructed with the actual data. Since the data is externally
+owned and already loaded into memory, this is beyond :func:`meta_init`'s
+control.
```

Comment:
Is this referring to `tensor.from_numpy`? If so, I'm wondering why this can't be intercepted since it is still a torch API?

Reply:
tensor.from_numpy never goes into the PyTorch dispatcher. If we wanted to intercept it, we would either have to do it at the `__torch_function__` level (i.e. at the level of PyTorch Python APIs) or figure out how to get the dispatcher to understand it

Reply:
As of today "operators" that construct tensors from external data cannot be intercepted by the dispatcher backends. This is an implementation limitation and there is ongoing discussion to fix it. Having said that, as I mentioned in the doc, I don't think that this is a problem in our case unless the user wants a copy of the data since the data is already in memory.